send unsolicited responses to a channel instead of discarding them

dario23 opened the pull request:
re #72 discussion

**notes**

  * i tried to avoid the term "async", because that term is very
    overloaded and we're not using e.g. tokio/async-io here
  * i'm a little unhappy having to string the channel through the
    parser, because that seems rather a part of the client logic than
    parsing. on the other hand it's better than passing the whole
    client, so there's that at least.

jonhoo commented:
This is great! Any particular reason you are cloning the `Sender` everywhere as opposed to passing `&mut self.unsolicited_responses_tx`?

dario23 commented:
> This is great! Any particular reason you are cloning the `Sender` everywhere as opposed to passing `&mut self.unsolicited_responses_tx`?

good point, changed that part, and fixed the syntax error from the previous commit :-)

jonhoo commented:
I'd also love to see a test that *does* test for an unsolicited response, as opposed to only ones that don't, if you can spare the time!

jonhoo commented:
Awesome, thanks! I also recently moved to `nom 4` in https://github.com/mattnenterprise/rust-imap/commit/5c91f4c1bbf8114641348382726230ae9b2e8553, so you'll have to rebase onto master (or a merge is also fine if you're not comfortable with git rebases) to make things work again. The change is mostly mechanical:
```rust
IResult::Done(rest, data)
```
just becomes
```rust
Ok((rest, data))
```

dario23 commented:
- added some tests, also some more handling along the way because tests found me being a little sloppy :-)
- rebased onto latest master (i like merges better when a branch was developed separately for some time and is later adopted into master, but for ongoing development like this i find linear history more aesthetic)

jonhoo commented:
Fantastic, thank you!
